ls1intum/Artemis
# `Plagiarism checks`: Increase interactivity of similarity distribution chart

### Checklist
#### General


#### Client

### Motivation and Context


This PR increase the interactivity of the similarity distribution chart of detected plagiarism cases

### Description

What is changed in this PR:

- **Help icon**: Next to the bar title is now a help icon with a tooltip explaining what is visualized by the chart and emphasizing that those cases have to be confirmed manually in order to punish them.
- **Enhanced bar labels**: The bar labels are now displaying the exact borders of the intervals in order to make it more transparent what is displayed by every chart bar
- **Color coding**: The bars representing cases with similarity >= 80% are now colored red
- **Datalabels**: Above each bar now a datalabel show the numeric value represented by the bar
- **Enhanced tooltips**: The tooltips of the bars are now more comprehensive and contain the relative amount they represent in regards to the number of total detected cases. In addition they itemize the represented cases in number of confirmed, declined and open cases.
- **Filter function**: By clicking on a bar, the cases displayed on the left side get filtered accordingly. Additionally, a disclaimer is shown informing the user that the cases are currently filtered and of course a reset button appears in the header.

Smaller additional improvements:

- The chart is now fitted to consume the whole page width
- The chart title is centered
- I refactored some code on different places. Mostly reoccurring functionality that is now implemented at a central place and reused

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Text/Modeling/Programming exercise with detected plagiarisms

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to the corresponding exercise detail page
4. Click "Check plagiarism"
5. If not done yet, run the plagiarism detection
6. Click "Run details" and inspect the chart. Make sure everything is comprehensive. Check that the tooltips itemize the cases correctly
7. Confirm a case. Check that the tooltip of the corresponding bar updated correctly
8. Do the same after denying a case
9. Select an arbitrary bar. Make sure the cases get filtered correctly and that the numbering is still as it is without filtering. Also make sure that a disclaimer appears after filtering as well as the reset button in the header.
10. Check that the confirming and denying still works when cases are filtered and the chart tooltips get updated accordingly. 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `plagiarism-inspector.component.ts` | 76.47 | 90.51|
| `plagiarism-inspector.service.ts` | 100 | 100 |
| `plagiarism-run-details.component.ts` | 77.77 | 87.5 |

### Screenshots


Example workflow:

https://user-images.githubusercontent.com/80622272/152644476-ca8c0111-6583-4c93-90d4-2c9195d50dcc.mp4
